features/telemetry/daemon/src/main/java/org/opennms/netmgt/telemetry/daemon/KarafStartupMonitor.java

```diff
@@ -0,0 +1,135 @@
+import org.opennms.netmgt.events.api.annotations.EventListener;
+import org.opennms.netmgt.xml.event.Event;
+import org.opennms.reporting.core.svclayer.ReportStoreService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.http.HttpEntity;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpMethod;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.client.RestTemplate;
+
+/**
+ * Used in startup script to detect if Karaf started properly.
+ * TODO: patrick move this class to better place
+ */
+@EventListener(name= KarafStartupMonitor.NAME, logPrefix= KarafStartupMonitor.LOG_PREFIX)
+public class KarafStartupMonitor implements SpringServiceDaemon {
+    private static final Logger LOG = LoggerFactory.getLogger(KarafStartupMonitor.class);
+
+    public static final String NAME = "KarafStartupMonitor";
+
+    public static final String LOG_PREFIX = "karafStartupMonitor";
+
+    @SuppressWarnings("unchecked")
+    private static final ServiceLookup<Class<?>, String> SERVICE_LOOKUP = new ServiceLookupBuilder(new ServiceRegistryLookup(DefaultServiceRegistry.INSTANCE))
+            .blocking()
+            .build();
+
+    @Override
+    public synchronized void start() {
+        LOG.info("{} is starting.", NAME);
+
+        if (!isKarafOk()) {
+        String message = String.format(NAME + ": It seems Kafka can't be started properly. This is bad, will fail startup.%n" +
+        "What can you do about this?%n" +
+        "1.) check in logs/karaf.log for problems%n" +
+        "2.) clear the 'data' folder - it contains Karaf's cache%n" +
+        "3.) run the script bin/fix-karaf-setup.sh");
+            throw new IllegalStateException(message);
+        }
+        LOG.info("{} is started.", NAME);
+    }
+
+    private boolean isKarafOk() {
+        Class serviceClass = ReportStoreService.class;
+        try {
+            final ReportStoreService service = SERVICE_LOOKUP.lookup(serviceClass, null);
+            return service != null;
+        } catch (Exception e) {
+            LOG.warn("Could not lookup {}, will fail.", serviceClass.getName(), e);
+            return false;
+        }
+    }
+
+    private boolean isKarafOkAlternative() {
+        try {
+            // TODO: if we use this method we wil hae to move the health service to the end of the list in org.apache.features.cfg
+            String healthEndpoint = "http://admin:admin@localhost:8980/opennms/rest/health"; // TODO: patrick Is this URL always correct?
+            String base64Creds = Base64.getEncoder().encodeToString("admin:admin".getBytes());
+            HttpHeaders headers = new HttpHeaders();
+            headers.add("Authorization", "Basic " + base64Creds);
+            HttpEntity request = new HttpEntity(headers);
+            ResponseEntity<String> response = new RestTemplate().exchange(healthEndpoint, HttpMethod.GET, request, String.class);
```

Comment:
This is not going to work well:
1. the URL is never the same. Should use `/opennms/rest/health` or configure username, password and url
2. Use /rest/health/probe instead, as it does not require authentication and simply returns 200 or > 500

Generally I would prefer to use the HealthCheckService here instead.
We could also add a custom HealthCheck, which verifies that ALL features defined in the features.boot.cfg have been installed correctly. Also if you cannot find the HealthCheckService, the container is not started properly

Reply:
I think we have two ways to check for Karaf health:

1. expose an http endpoint and query for it, done in isKarafOkAlternative() or
2. call a Karaf service directly, done in isKarafOk()
I think option 2 is the preferred one, once we get it to work properly we can remove isKarafOkAlternative().

Reply:
we removed the method